src/lib/form/cc-input-element.js

```diff
@@ -0,0 +1,363 @@
+    this.name = null;
+  }
+
+  /**
+   * @return {InputElementSettings}
+   * @protected
+   */
+  getInputSettings () {
+    throw new Error('You must implement getInputSettings() method!');
+  }
+
+  /**
+   * This callback will be called when the form associated to the element is reset.
+   */
+  formResetCallback () {
+    this._formHelper.resetValue();
+    // we really need to reset the error message because, even if input becomes invalid after reset,
+    // the form reset must also reset the error message.
+    this.validate(false);
+    if (this.errorMessage != null) {
+      this.errorMessage = null;
+      dispatchCustomEvent(this, 'error-message-change', null);
+    }
+  }
+
+  /**
+   * @param {boolean} report - whether to display error message or not
+   */
+  validate (report) {
+    const validationSettings = this._formHelper.getValidationSettings();
+
+    /** @type {ErrorMessageMap} */
+    const errorMessages = {
+      ...validationSettings.errorMessages,
+      ...(this.customErrorMessages ?? {}),
+    };
+
+    const validator = CompositeValidator.builder()
+      .add(validationSettings.validator)
+      .add(this.customValidator)
+      .build();
+
+    const validation = validator.validate(
+      this._formHelper.getValueProperty(),
+      this._formHelper.internals.form != null ? getFormData(this._formHelper.internals.form) : {},
+    );
+
+    const errorMessage = validation.valid === true
+      ? null
+      : this._formHelper.resolveErrorMessage(validation.code, errorMessages);
+
+    if (validation.valid) {
+      this._formHelper.setValidValidity();
+    }
+    else {
+      this._formHelper.setInvalidValidity(convertErrorMessageToString(errorMessage));
+    }
+
+    if (report) {
+      if (errorMessage !== this.errorMessage) {
+        this.errorMessage = errorMessage;
+        dispatchCustomEvent(this, 'error-message-change', errorMessage);
+      }
+    }
+
+    return validation;
+  }
+
+  /* region mimic the native validation API */
+
+  /**
+   * Returns whether an element will successfully validate based on forms validation rules and constraints.
+   * {@link https://developer.mozilla.org/en-US/docs/Web/API/ElementInternals/willValidate}
+   *
+   * @return {boolean}
+   */
+  get willValidate () {
+    return this._formHelper.internals.willValidate;
+  }
+
+  /**
+   * Mimics the native HTMLInputElement API.
+   * {@link https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement/checkValidity}
+   *
+   * @return {boolean}
+   */
+  checkValidity () {
+    return this._formHelper.internals.checkValidity();
+  }
+
+  /**
+   * Mimics the native HTMLInputElement API.
+   * {@link https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement/reportValidity}
+   *
+   * @return {HTMLInputElement}
+   */
+  reportValidity () {
+    return this._formHelper.internals.reportValidity();
+  }
+
+  /**
+   * Returns a `ValidityState` object that represents the validity states of an element.
```

Comment:
An interface is a definition. Here it returns an object (I could have said an _instance_) that meets an interface contract.

Reply:
You're right sorry, I like instance, I was nitpicking because I just wanted people to know that they wouldn't get a standard Object (for instance all of its properties are readonly).